File: 46. 全排列.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [[]]
        result = []
        for i in range(len(nums)):
            first = nums[i]
            rest = nums[:i] + nums[i + 1:]
            permutations = self.permute(rest)
            for permutation in permutations:
                result.append([first] + permutation)
        return result


solution = Solution()
print(solution.permute([1, 2, 3]))

File: 46. 全排列.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    if (nums.length === 0) return []
    let result = []
    let backtracking = (nums, path) => {
        if (path.length === nums.length) {
            result.push(path.slice());
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (path.includes(nums[i])) {
                continue;
            }
            path.push(nums[i])
            backtracking(nums, path)
            path.pop()
        }
    }
    backtracking(nums, [])
    return result
};
let nums = [1, 2, 3]
console.log(permute(nums));